src/types.py

```diff
@@ -1049,6 +1054,61 @@ def try_add_field(
         self.fields.sort(key=lambda f: f.offset)
         return field
 
+    def prune_overlapping_fields(self) -> None:
+        """Remove overlapping fields with `known=False` from the struct"""
+
+        # TODO: Support unions
+        if self.is_union:
+            return None
+
+        # Sort by offset, with bigger fields at the same offset first
+        self.fields.sort(key=lambda f: (f.offset, -(f.type.get_size_bytes() or 0)))
+
+        fields_to_remove: Set[StructDeclaration.StructField] = set()
+        for i, field in enumerate(self.fields):
+            # Skip fields provided by the context or marked for removal
+            if field.known or field in fields_to_remove:
+                continue
+
+            # If the size of a field (still) isn't known, assume it is 1 byte
+            field_size = field.type.get_size_bytes() or 1
+
+            conflicting_fields: Set[StructDeclaration.StructField] = set()
+            for f2 in self.fields[i + 1 :]:
+                assert f2.offset >= field.offset
+                # If `f2` is after the end of `field`, we're done
+                if f2.offset >= field.offset + field_size:
+                    break
+                if f2 in fields_to_remove:
+                    continue
+
+                # Now, we know `field` and `f2` overlap. Check if `f2` is a subfield of `field`.
+                offset = f2.offset - field.offset
+                sub_path, sub_type, _ = field.type.get_field(
+                    offset, target_size=f2.type.get_size_bytes()
+                )
+                if sub_path is not None and sub_type.unify(f2.type):
+                    # Remove `f2` so that accesses to `f2.offset` go through `field` instead.
+                    # Defer adding `f2` to `fields_to_remove` until we're done processing `field`.
+                    conflicting_fields.add(f2)
+                elif not f2.known and f2.type.is_int() and field.type.is_int():
+                    # Ignore overlapping inferred ints.
+                    # Usually the code is easier to manually fix if we leave the overlap
+                    pass
```

Comment:
I could remove this `elif`, there are just 2 cases in MM where it matters (`audio_playback` and `pfsdeletefile`). It's not super clear to me if it appreciably helps, or is worth the complexity?

Reply:
Seems fine, worth extending the comment to explain that this is because of load+cast being converted to load of smaller size. Eventually it would be nice to start detecting such loads, I assume this code can go away at that point